pex/bin/pex.py

```diff
@@ -291,6 +291,17 @@ def configure_clp_pex_options(parser):
            'No value (alias for prefer, for backwards compatibility). '
            '[Default: %default]')
 
+  group.add_option(
+      '--reproducible', '--not-reproducible',
```

Comment:
I have not convinced myself that we want to call the flag this, specifically because of timestamps. 

Due to the way zipfiles work, we cannot have a reproducible build without somehow hardcoding the timestamp. Per  https://reproducible-builds.org/docs/source-date-epoch/, the canonical way to do this is via an env var `SOURCE_DATE_EPOCH`. If we don't use this env var, then we would have to ourselves hardcode some arbitrary value for what timestamp the zip file should use, which seems very wrong.

So, my hesitation is that unless you specify `SOURCE_DATE_EPOCH`, the build will _not_ be truly reproducible. I don't want the flag to make a false promise.

--

Just thought of a good workaround. If they specify `--reproducible`, then we eagerly check that `SOURCE_DATE_EPOCH` is set in their env vars. If not, we fail and request they set this, then run again.

I advocate for this approach.

--

If we don't go with this workaround and don't like `--reproducible`, then we could rename this to something like `--pyc-files-included` / `--no-pyc-files-included` or `--byte-code-included` / `--no-byte-code-included`.

Reply:
@jsirois 

> As to names for things - I'm never right. I'll leave that debate to you all.

Fair, but could you weigh in please on the `SOURCE_DATE_EPOCH` idea of them having to supply it via env var vs. us hardcoding an arbitrary value? If we indeed are going to ask them to `SOURCE_DATE_EPOCH`, which per Reproducible Builds site is the canonical way to do this, then I think both decisions are easy to make.

Reply:
I think things that are necessary for reproducible builds should be specified in options scoped specifically to reproducible builds. If a value named SOURCE_DATE_EPOCH is required, we could make an option `--reproducible-source-date-epoch=<epoch>` which would be required if `--reproducible-build=<output pex file name>` was provided, and otherwise ignored.

Reply:
> Fair, but could you weigh in please on the SOURCE_DATE_EPOCH idea...

Honoring `SOURCE_DATE_EPOCH` sounds sane, forcing it to be specified does not make sense to me. Why can't we use 0 as the default? It may be arbitrary, but that's the whole point - the value is arbitrary.

Reply:
> Keeping this as a flag that modifies -o, rather than introducing a new command, fits in with how we currently approach --zip-safe, --always-write-cache, --ignore-errors, and --inherit-path.

Ok, that sounds a lot more correct -- let's do that.